### PR TITLE
Added new oscillator types and improved its handling

### PR DESCRIPTION
**Pull request by uis246**
Not tested myself

### CODE REVIEW

**Comment by uis246** [comment]
Some idiotic mistakes were made. Now waiting again.

**Comment by uis246** [comment]
Looks promising. I'm waiting for #164 first.

File: apycula/gowin_unpack.py [hunk 1]
```diff
@@ -610,11 +637,13 @@ def tile2verilog(dbrow, dbcol, bels, pips, clock_pips, mod, cst, db):
...
             mod.wires.update(chain.from_iterable([x if isinstance(x, list) else [x] for x in ram16.portmap.values()]))
             mod.primitives[name] = ram16
-        elif typ in {"OSC", "OSCZ", "OSCF", "OSCH"}:
+        elif typ in {"OSC", "OSCZ", "OSCF", "OSCH", "OSCW", "OSCO"}:
             name = f"R{row}C{col}_{typ}"
             osc = codegen.Primitive(typ, name)
-            divisor, = flags
-            osc.params["FREQ_DIV"] = f"{divisor*2}"
+            for paramval in flags:
+                print(paramval)
```
Comment (uis246) [inline apycula/gowin_unpack.py:644]: Ooops

File: apycula/gowin_pack.py [hunk 2]
```diff
@@ -359,15 +385,14 @@ def place(db, tilemap, bels, cst, args):
...
-            if divisor in tiledata.bels[typ].modes:
-                bits = tiledata.bels[typ].modes[divisor]
-                for r, c in bits:
-                    tile[r][c] = 1
+        elif typ in {'OSC', 'OSCZ', 'OSCF', 'OSCH', 'OSCW', 'OSCO'}:
+            osc_attrs = set_osc_attrs(db, typ, parms)
+            bits = set()
+            if 'OSC' in db.shortval[tiledata.ttyp].keys():
+                bits = get_shortval_fuses(db, tiledata.ttyp, osc_attrs, 'OSC')
+            print(bits)
```
Comment (uis246) [inline apycula/gowin_pack.py:393]: And here

**Comment by uis246** [comment]
I think it's better to merge it as single commit

**Comment by pepijndevos** [comment]
Awesome. I'll review it next week when I get home. Or maybe yrabbit will have a look. Really exciting to see your contributions!

**Comment by yrabbit** [comment]
> Awesome. I'll review it next week when I get home. Or maybe yrabbit will have a look. Really exciting to see your contributions!

With my eyes it seems fine, but I don't have any GW2 series boards (it's a Tang Primer 20K, right?) so it's hard to check  on hw.

**Comment by yrabbit** [comment]
Take the new master and fix your rep, please.

**Comment by yrabbit** [comment]
tangnano9k and tangnano4k - no questions asked, binky-osc works.

tangnano (GW1N-1) generator does not start, here is the movie (first the usual blinky, then blinky-osc-tangnano.fs is loaded) and the output file after nextpnr and binary image. If you can't figure it out by Saturday, I'll look into it:)

https://user-images.githubusercontent.com/6075465/223281505-7d424f26-add0-466f-ad63-a93b295a7355.mp4

[blinky-osc-tangnano.fs.gz](https://github.com/YosysHQ/apicula/files/10903873/blinky-osc-tangnano.fs.gz)
[blinky-osc-tangnano.json.gz](https://github.com/YosysHQ/apicula/files/10903874/blinky-osc-tangnano.json.gz)

**Comment by yrabbit** [comment]
'POWER_SAVE' = 'ENABLE' 
The set fuse 2204 puts the generator into energy saving mode effectively stopping it. 

There are different ways to correct this, and it makes sense to check the other supported chips and at least take this parameter into account :)

powersave

**Comment by uis246** [comment]
What if you set 'NORMAL' = 'ENABLE'?

**Comment by yrabbit** [comment]
> What if you set 'NORMAL' = 'ENABLE'? Or this is set anyways when OSC is present because default FREQ_DIV is still FREQ_DIV

No, the fuses of this attribute are set by default in the "empty" image, which is then filled in gowin_pack.

That's why I'm saying that there are all sorts of options here to fix this: either do something with the "empty" image or provide for cleanup in gowin_pack...

**Comment by yrabbit** [comment]
In the attached file, some edits, after which OSC will work on Tangnano, Tangnano1k, Tangnano4k, Tangnano9k. 
* Take into account the fact that some inputs and outputs are in other cells from the primitive
* Clear energy conservation bits inherited from image template
* Enable OSC in the GW1NZ-1 chip by setting the impirical bits.
* Minor edits.

[0.txt](https://github.com/YosysHQ/apicula/files/11261243/0.txt)

**Comment by yrabbit** [comment]
Included in   https://github.com/YosysHQ/apicula/pull/174

**Pull request closed without merging**